File: src/dfs/dfs.py
class Node:
    def __init__(self, value, adj):
        self.value = value
        self.adj = adj
        self.visited = False

def dfs(graph):
    for source in graph:
        source.visited = True
        dfs_node(source)

def dfs_node(node):
    print(node.value)
    for e in node.adj:
        if not e.visited:
            e.visited = True
            dfs_node(e)

File: src/dfs/dfs.hs
import Data.List

data Node = Node { value :: Int, adj :: [Node] } deriving Eq

instance Show Node where
    show node = show $ value node

-- source node의 집합
type Graph = [Node]

dfs :: Graph -> [Node]
dfs graph = concat [dfs_node x [] | x <-graph]
    where dfs_node source visited = source:(concat [dfs_node a (source:visited) | a <- adj source, a `notElem` visited])